{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b450b64a",
   "metadata": {},
   "source": [
    "note : we need all the genes to do the DE analysis yaaay \n",
    "but, i found a way to load the whole file (it was a compressing issue when creating the file after reducing the dimensions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e8829c0",
   "metadata": {},
   "source": [
    "set the random seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "be15ea8a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "the.seed <-42\n",
    "set.seed(the.seed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09534073",
   "metadata": {},
   "source": [
    "packages installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1607d770",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//RtmpG98ux2/downloaded_packages\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//RtmpG98ux2/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GitHub PAT from the git credential store.\n",
      "\n",
      "Skipping install of 'SeuratDisk' from a github remote, the SHA1 (877d4e18) has not changed since last install.\n",
      "  Use `force = TRUE` to force installation\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//RtmpG98ux2/downloaded_packages\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//RtmpG98ux2/downloaded_packages\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//RtmpG98ux2/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.21 (BiocManager 1.30.27), R 4.5.1 (2025-06-13)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'zellkonverter' 'HDF5Array'”\n",
      "Old packages: 'downlit', 'fansi', 'gert', 'GetoptLong', 'ggbeeswarm',\n",
      "  'ggtangle', 'GlobalOptions', 'gprofiler2', 'lintr', 'mgcv', 'pkgdown',\n",
      "  'RSQLite', 'testthat', 'XML', 'xml2', 'yaml'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.21 (BiocManager 1.30.27), R 4.5.1 (2025-06-13)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'muscat'”\n",
      "Old packages: 'downlit', 'fansi', 'gert', 'GetoptLong', 'ggbeeswarm',\n",
      "  'ggtangle', 'GlobalOptions', 'gprofiler2', 'lintr', 'mgcv', 'pkgdown',\n",
      "  'RSQLite', 'testthat', 'XML', 'xml2', 'yaml'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"Seurat\")\n",
    "install.packages(\"remotes\")\n",
    "remotes::install_github(\"mojaveazure/seurat-disk\")\n",
    "install.packages(\"hdf5r\")\n",
    "install.packages(\"reticulate\")\n",
    "install.packages(\"BiocManager\")\n",
    "BiocManager::install(c(\"zellkonverter\", \"HDF5Array\"))\n",
    "BiocManager::install(\"muscat\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f718dd7",
   "metadata": {},
   "source": [
    "library loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a9e05f01",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'zellkonverter':\n",
      "  method                                             from      \n",
      "  py_to_r.pandas.core.arrays.categorical.Categorical reticulate\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.5.0 but the current version is\n",
      "4.5.1; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "Warning message:\n",
      "“package ‘reticulate’ was built under R version 4.5.2”\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: generics\n",
      "\n",
      "\n",
      "Attaching package: ‘generics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratDisk’:\n",
      "\n",
      "    as.factor\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
      "    setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
      "    unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:generics’:\n",
      "\n",
      "    explain\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Warning message in check_dep_version(dep_pkg = \"TMB\"):\n",
      "“package version mismatch: \n",
      "glmmTMB was built with TMB package version 1.9.17\n",
      "Current TMB package version is 1.9.18\n",
      "Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)”\n"
     ]
    }
   ],
   "source": [
    "\n",
    "library(zellkonverter)\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(zellkonverter)\n",
    "library(reticulate)\n",
    "library(SingleCellExperiment)\n",
    "library(dplyr)\n",
    "library(muscat)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44f7657d",
   "metadata": {},
   "source": [
    "choose path for data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a8f75981",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# elodie's path : \n",
    "path = \"/Users/elodiehusson/Desktop/AD & PD/\"\n",
    "\n",
    "#Yara's path :\n",
    "# path = \"C:/Users/yarad/Desktop/x/Masters/Master BMC - Sorbonne/M2/Single Cell/Project/Coding Project/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "452b2e37",
   "metadata": {},
   "source": [
    "i use a single cell experiment object. It's not a Seurat object (too complicated !!) - maybe we can use a seurat object but it seems like it's complicated. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4c9ff384",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "anndata._core.sparse_dataset._CSRDatasetanndata._core.sparse_dataset.BaseCompressedSparseDatasetanndata.abc.CSRDatasetanndata.abc._AbstractCSDatasetabc.ABCpython.builtin.object\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: SingleCellExperiment \n",
       "dim: 34176 77052 \n",
       "metadata(12): batch_condition citation ... title uid\n",
       "assays(1): X\n",
       "rownames(34176): ENSG00000000003 ENSG00000000005 ... ENSG00000288642\n",
       "  ENSG00000288643\n",
       "rowData names(8): gene_name n_cells ... feature_length feature_type\n",
       "colnames(77052): Donor_333-1-GTCTGTCGTTCTCCTG-1\n",
       "  Donor_1001-1-ACTTTGTGTTACCGTA-1 ... Donor_634-2-ACGATCATCATGAAAG-0\n",
       "  Donor_721-1-AAGCGAGCAGCGTACC-0\n",
       "colData names(36): Source n_genes ... development_stage\n",
       "  observation_joinid\n",
       "reducedDimNames(1): X_umap\n",
       "mainExpName: NULL\n",
       "altExpNames(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sce <- readH5AD(\n",
    "    file.path(path, \"77k_AD_PD_CTRL.h5ad\"), #loading the file with our paths\n",
    "    use_hdf5 = TRUE, # reduces part of what is loaded in memory (to reduce RAM usage, and time)\n",
    ")\n",
    "sce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "88b6ac7d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<11 x 11> sparse DelayedMatrix object of type \"double\":\n",
       "                 Donor_865-1-TCGCACTCAGGACTTT-0 ...\n",
       "ENSG00000005801                               2   .\n",
       "ENSG00000005810                              33   .\n",
       "ENSG00000005812                               6   .\n",
       "ENSG00000005844                               2   .\n",
       "ENSG00000005882                               0   .\n",
       "ENSG00000005884                               1   .\n",
       "ENSG00000005889                               2   .\n",
       "ENSG00000005893                               3   .\n",
       "ENSG00000005961                               0   .\n",
       "ENSG00000005981                               0   .\n",
       "ENSG00000006007                               1   .\n",
       "                Donor_1001-1-TTTCCTCGTTGCATCA-1\n",
       "ENSG00000005801                               0\n",
       "ENSG00000005810                               4\n",
       "ENSG00000005812                               2\n",
       "ENSG00000005844                               0\n",
       "ENSG00000005882                               0\n",
       "ENSG00000005884                               0\n",
       "ENSG00000005889                               0\n",
       "ENSG00000005893                               0\n",
       "ENSG00000005961                               0\n",
       "ENSG00000005981                               0\n",
       "ENSG00000006007                               0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 11 × 11 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Donor_865-1-TCGCACTCAGGACTTT-0</th><th scope=col>Donor_1001-1-TAGGGTTCACGGCTAC-1</th><th scope=col>Donor_865-1-TACGCTCCACTGAATC-0</th><th scope=col>Donor_333-1-GATGATCCATGCGGTC-1</th><th scope=col>Donor_865-2-GTCACGGCACCAGACC-1</th><th scope=col>Donor_865-1-ATAGAGAAGGCTTAGG-0</th><th scope=col>Donor_865-2-TCGCAGGAGCACACAG-1</th><th scope=col>Donor_31-1-CGACAGCGTGCAGTGA-0</th><th scope=col>Donor_865-1-GGTTGTAAGCGTCTGC-0</th><th scope=col>Donor_31-1-AATCACGGTGACTAAA-0</th><th scope=col>Donor_1001-1-TTTCCTCGTTGCATCA-1</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000005801</th><td> 2</td><td>0</td><td>0</td><td> 2</td><td>0</td><td>1</td><td> 2</td><td> 2</td><td>0</td><td> 2</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005810</th><td>33</td><td>2</td><td>0</td><td>15</td><td>8</td><td>1</td><td>15</td><td>16</td><td>0</td><td>14</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005812</th><td> 6</td><td>0</td><td>1</td><td> 2</td><td>0</td><td>0</td><td> 1</td><td> 1</td><td>1</td><td> 3</td><td>2</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005844</th><td> 2</td><td>0</td><td>0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005882</th><td> 0</td><td>1</td><td>0</td><td> 1</td><td>0</td><td>0</td><td> 0</td><td> 0</td><td>0</td><td> 1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005884</th><td> 1</td><td>0</td><td>0</td><td> 2</td><td>0</td><td>0</td><td> 0</td><td> 1</td><td>0</td><td> 3</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005889</th><td> 2</td><td>0</td><td>0</td><td> 2</td><td>0</td><td>0</td><td> 3</td><td> 2</td><td>0</td><td> 1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005893</th><td> 3</td><td>0</td><td>9</td><td> 7</td><td>0</td><td>6</td><td> 1</td><td> 0</td><td>3</td><td> 2</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005961</th><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td> 0</td><td>1</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000005981</th><td> 0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td>0</td><td> 0</td><td> 0</td><td>0</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000006007</th><td> 1</td><td>0</td><td>0</td><td> 1</td><td>0</td><td>0</td><td> 0</td><td> 1</td><td>0</td><td> 1</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 11 × 11 of type dbl\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Donor\\_865-1-TCGCACTCAGGACTTT-0 & Donor\\_1001-1-TAGGGTTCACGGCTAC-1 & Donor\\_865-1-TACGCTCCACTGAATC-0 & Donor\\_333-1-GATGATCCATGCGGTC-1 & Donor\\_865-2-GTCACGGCACCAGACC-1 & Donor\\_865-1-ATAGAGAAGGCTTAGG-0 & Donor\\_865-2-TCGCAGGAGCACACAG-1 & Donor\\_31-1-CGACAGCGTGCAGTGA-0 & Donor\\_865-1-GGTTGTAAGCGTCTGC-0 & Donor\\_31-1-AATCACGGTGACTAAA-0 & Donor\\_1001-1-TTTCCTCGTTGCATCA-1\\\\\n",
       "\\hline\n",
       "\tENSG00000005801 &  2 & 0 & 0 &  2 & 0 & 1 &  2 &  2 & 0 &  2 & 0\\\\\n",
       "\tENSG00000005810 & 33 & 2 & 0 & 15 & 8 & 1 & 15 & 16 & 0 & 14 & 4\\\\\n",
       "\tENSG00000005812 &  6 & 0 & 1 &  2 & 0 & 0 &  1 &  1 & 1 &  3 & 2\\\\\n",
       "\tENSG00000005844 &  2 & 0 & 0 &  0 & 0 & 0 &  0 &  0 & 0 &  0 & 0\\\\\n",
       "\tENSG00000005882 &  0 & 1 & 0 &  1 & 0 & 0 &  0 &  0 & 0 &  1 & 0\\\\\n",
       "\tENSG00000005884 &  1 & 0 & 0 &  2 & 0 & 0 &  0 &  1 & 0 &  3 & 0\\\\\n",
       "\tENSG00000005889 &  2 & 0 & 0 &  2 & 0 & 0 &  3 &  2 & 0 &  1 & 0\\\\\n",
       "\tENSG00000005893 &  3 & 0 & 9 &  7 & 0 & 6 &  1 &  0 & 3 &  2 & 0\\\\\n",
       "\tENSG00000005961 &  0 & 0 & 0 &  0 & 0 & 0 &  0 &  0 & 1 &  0 & 0\\\\\n",
       "\tENSG00000005981 &  0 & 0 & 0 &  0 & 0 & 0 &  0 &  0 & 0 &  0 & 0\\\\\n",
       "\tENSG00000006007 &  1 & 0 & 0 &  1 & 0 & 0 &  0 &  1 & 0 &  1 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 11 × 11 of type dbl\n",
       "\n",
       "| <!--/--> | Donor_865-1-TCGCACTCAGGACTTT-0 | Donor_1001-1-TAGGGTTCACGGCTAC-1 | Donor_865-1-TACGCTCCACTGAATC-0 | Donor_333-1-GATGATCCATGCGGTC-1 | Donor_865-2-GTCACGGCACCAGACC-1 | Donor_865-1-ATAGAGAAGGCTTAGG-0 | Donor_865-2-TCGCAGGAGCACACAG-1 | Donor_31-1-CGACAGCGTGCAGTGA-0 | Donor_865-1-GGTTGTAAGCGTCTGC-0 | Donor_31-1-AATCACGGTGACTAAA-0 | Donor_1001-1-TTTCCTCGTTGCATCA-1 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000005801 |  2 | 0 | 0 |  2 | 0 | 1 |  2 |  2 | 0 |  2 | 0 |\n",
       "| ENSG00000005810 | 33 | 2 | 0 | 15 | 8 | 1 | 15 | 16 | 0 | 14 | 4 |\n",
       "| ENSG00000005812 |  6 | 0 | 1 |  2 | 0 | 0 |  1 |  1 | 1 |  3 | 2 |\n",
       "| ENSG00000005844 |  2 | 0 | 0 |  0 | 0 | 0 |  0 |  0 | 0 |  0 | 0 |\n",
       "| ENSG00000005882 |  0 | 1 | 0 |  1 | 0 | 0 |  0 |  0 | 0 |  1 | 0 |\n",
       "| ENSG00000005884 |  1 | 0 | 0 |  2 | 0 | 0 |  0 |  1 | 0 |  3 | 0 |\n",
       "| ENSG00000005889 |  2 | 0 | 0 |  2 | 0 | 0 |  3 |  2 | 0 |  1 | 0 |\n",
       "| ENSG00000005893 |  3 | 0 | 9 |  7 | 0 | 6 |  1 |  0 | 3 |  2 | 0 |\n",
       "| ENSG00000005961 |  0 | 0 | 0 |  0 | 0 | 0 |  0 |  0 | 1 |  0 | 0 |\n",
       "| ENSG00000005981 |  0 | 0 | 0 |  0 | 0 | 0 |  0 |  0 | 0 |  0 | 0 |\n",
       "| ENSG00000006007 |  1 | 0 | 0 |  1 | 0 | 0 |  0 |  1 | 0 |  1 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "                Donor_865-1-TCGCACTCAGGACTTT-0 Donor_1001-1-TAGGGTTCACGGCTAC-1\n",
       "ENSG00000005801  2                             0                              \n",
       "ENSG00000005810 33                             2                              \n",
       "ENSG00000005812  6                             0                              \n",
       "ENSG00000005844  2                             0                              \n",
       "ENSG00000005882  0                             1                              \n",
       "ENSG00000005884  1                             0                              \n",
       "ENSG00000005889  2                             0                              \n",
       "ENSG00000005893  3                             0                              \n",
       "ENSG00000005961  0                             0                              \n",
       "ENSG00000005981  0                             0                              \n",
       "ENSG00000006007  1                             0                              \n",
       "                Donor_865-1-TACGCTCCACTGAATC-0 Donor_333-1-GATGATCCATGCGGTC-1\n",
       "ENSG00000005801 0                               2                            \n",
       "ENSG00000005810 0                              15                            \n",
       "ENSG00000005812 1                               2                            \n",
       "ENSG00000005844 0                               0                            \n",
       "ENSG00000005882 0                               1                            \n",
       "ENSG00000005884 0                               2                            \n",
       "ENSG00000005889 0                               2                            \n",
       "ENSG00000005893 9                               7                            \n",
       "ENSG00000005961 0                               0                            \n",
       "ENSG00000005981 0                               0                            \n",
       "ENSG00000006007 0                               1                            \n",
       "                Donor_865-2-GTCACGGCACCAGACC-1 Donor_865-1-ATAGAGAAGGCTTAGG-0\n",
       "ENSG00000005801 0                              1                             \n",
       "ENSG00000005810 8                              1                             \n",
       "ENSG00000005812 0                              0                             \n",
       "ENSG00000005844 0                              0                             \n",
       "ENSG00000005882 0                              0                             \n",
       "ENSG00000005884 0                              0                             \n",
       "ENSG00000005889 0                              0                             \n",
       "ENSG00000005893 0                              6                             \n",
       "ENSG00000005961 0                              0                             \n",
       "ENSG00000005981 0                              0                             \n",
       "ENSG00000006007 0                              0                             \n",
       "                Donor_865-2-TCGCAGGAGCACACAG-1 Donor_31-1-CGACAGCGTGCAGTGA-0\n",
       "ENSG00000005801  2                              2                           \n",
       "ENSG00000005810 15                             16                           \n",
       "ENSG00000005812  1                              1                           \n",
       "ENSG00000005844  0                              0                           \n",
       "ENSG00000005882  0                              0                           \n",
       "ENSG00000005884  0                              1                           \n",
       "ENSG00000005889  3                              2                           \n",
       "ENSG00000005893  1                              0                           \n",
       "ENSG00000005961  0                              0                           \n",
       "ENSG00000005981  0                              0                           \n",
       "ENSG00000006007  0                              1                           \n",
       "                Donor_865-1-GGTTGTAAGCGTCTGC-0 Donor_31-1-AATCACGGTGACTAAA-0\n",
       "ENSG00000005801 0                               2                           \n",
       "ENSG00000005810 0                              14                           \n",
       "ENSG00000005812 1                               3                           \n",
       "ENSG00000005844 0                               0                           \n",
       "ENSG00000005882 0                               1                           \n",
       "ENSG00000005884 0                               3                           \n",
       "ENSG00000005889 0                               1                           \n",
       "ENSG00000005893 3                               2                           \n",
       "ENSG00000005961 1                               0                           \n",
       "ENSG00000005981 0                               0                           \n",
       "ENSG00000006007 0                               1                           \n",
       "                Donor_1001-1-TTTCCTCGTTGCATCA-1\n",
       "ENSG00000005801 0                              \n",
       "ENSG00000005810 4                              \n",
       "ENSG00000005812 2                              \n",
       "ENSG00000005844 0                              \n",
       "ENSG00000005882 0                              \n",
       "ENSG00000005884 0                              \n",
       "ENSG00000005889 0                              \n",
       "ENSG00000005893 0                              \n",
       "ENSG00000005961 0                              \n",
       "ENSG00000005981 0                              \n",
       "ENSG00000006007 0                              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x <- assay(sce, \"X\")[100:110, 100:110]\n",
    "x\n",
    "as.matrix(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5745dea8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "List of length 1\n",
       "names(1): X"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# gives the normalized matrix of counts\n",
    "assays(sce) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e7aa06c3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "logcounts(sce) <- assay(sce, \"X\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "111625a0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'X'</li><li>'logcounts'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'X'\n",
       "\\item 'logcounts'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'X'\n",
       "2. 'logcounts'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"X\"         \"logcounts\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assayNames(sce)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0c28a3",
   "metadata": {},
   "source": [
    "CODE CHATGPT POUR EXPRESSION DIFFERENTIELLE : D'ABORD, ON S'OCCUPE D'ORGANISER LES DONNEES DANS LE SCE\n",
    "mais du coup chatgpt m'a dit que on pouvait pas faire de DESeq2, ou edgeR pq les données sont normalisées, jsp si c'est vrai. \n",
    "aussi, une autre option est de transformer le SCE en un objet Seurat, et puis continuer la DE avec. Mais là jsplus quoi faire. \n",
    "je l'ai run, et c'est très long (j'ai arreté, pq ca voulait pas s'arreter de tourner), donc jsp si c'est normal ou quoi, mais c'est le plus loin que j'ai pu aller."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1380a007",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### 1) Harmoniser les colonnes du SCE \n",
    "\n",
    "# group_id = la condition d'intérêt\n",
    "sce$group_id <- case_when(\n",
    "    sce$disease == \"dementia || Alzheimer disease\" ~ \"AD\",\n",
    "    sce$disease == \"dementia || Parkinson disease\" ~ \"PD\",\n",
    "    sce$disease == \"normal\" ~ \"CTRL\",\n",
    "    TRUE ~ NA_character_\n",
    ")\n",
    "\n",
    "# sample_id = donneur\n",
    "sce$sample_id <- as.character(sce$donor_id)\n",
    "\n",
    "# cluster_id = type cellulaire\n",
    "# (tu peux aussi choisir sce$subclass ou sce$subtype selon ta résolution)\n",
    "sce$cluster_id <- as.character(sce$cell_type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62a385d3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "   AD  CTRL    PD \n",
       "31949 37331  7772 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "                                                     astrocyte \n",
       "                                                          9235 \n",
       "                                                        B cell \n",
       "                                                             3 \n",
       "                                              endothelial cell \n",
       "                                                           740 \n",
       "                                              GABAergic neuron \n",
       "                                                          1183 \n",
       "       L2/3 intratelencephalic projecting glutamatergic neuron \n",
       "                                                          9897 \n",
       "     L2/3-6 intratelencephalic projecting glutamatergic neuron \n",
       "                                                          5546 \n",
       "L5 extratelencephalic projecting glutamatergic cortical neuron \n",
       "                                                            12 \n",
       "                     L5/6 near-projecting glutamatergic neuron \n",
       "                                                           378 \n",
       "   L6 corticothalamic-projecting glutamatergic cortical neuron \n",
       "                                                           652 \n",
       "         L6 intratelencephalic projecting glutamatergic neuron \n",
       "                                                          1177 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### 2) On ne garde que AD / PD / CTRL \n",
    "sce <- sce[, !is.na(sce$group_id)]\n",
    "\n",
    "\n",
    "### 3) Vérifier que les colonnes nécessaires existent bien \n",
    "stopifnot(\n",
    "  \"sample_id\"  %in% colnames(colData(sce)),\n",
    "  \"group_id\"   %in% colnames(colData(sce)),\n",
    "  \"cluster_id\" %in% colnames(colData(sce))\n",
    ")\n",
    "\n",
    "\n",
    "### 4) S'assurer que tout est en character \n",
    "sce$sample_id  <- as.character(sce$sample_id)\n",
    "sce$group_id   <- as.character(sce$group_id)\n",
    "sce$cluster_id <- as.character(sce$cluster_id)\n",
    "\n",
    "\n",
    "### 5) Vérif simple\n",
    "table(sce$group_id)\n",
    "table(sce$cluster_id)[1:10]  # affichage partiel\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "67aef090",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "### 8) Pseudobulk : utiliser ASSAY = \"X\" ----\n",
    "# car ton fichier n'a PAS de \"logcounts\", seulement \"X\"\n",
    "\n",
    "pb <- aggregateData(\n",
    "  sce,\n",
    "  assay = \"X\",       # <- LE BON CHOIX\n",
    "  fun   = \"mean\",    # pour log-expressions\n",
    "  by    = c(\"cluster_id\", \"sample_id\")\n",
    ")\n",
    "\n",
    "pb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0fcb3bdd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown file type: h5ad”\n",
      "Creating h5Seurat file for version 3.1.5.9900\n",
      "\n",
      "Adding X as data\n",
      "\n",
      "Adding X as counts\n",
      "\n",
      "Adding meta.features from var\n",
      "\n",
      "Adding X_umap as cell embeddings for umap\n",
      "\n",
      "Adding batch_condition to miscellaneous data\n",
      "\n",
      "Adding citation to miscellaneous data\n",
      "\n",
      "Adding class_colors to miscellaneous data\n",
      "\n",
      "Adding genome to miscellaneous data\n",
      "\n",
      "Adding organism to miscellaneous data\n",
      "\n",
      "Adding organism_ontology_term_id to miscellaneous data\n",
      "\n",
      "Adding schema_reference to miscellaneous data\n",
      "\n",
      "Adding schema_version to miscellaneous data\n",
      "\n",
      "Adding subclass_colors to miscellaneous data\n",
      "\n",
      "Adding subtype_colors to miscellaneous data\n",
      "\n",
      "Adding title to miscellaneous data\n",
      "\n",
      "Adding uid to miscellaneous data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SeuratDisk)\n",
    "\n",
    "Convert(\n",
    "    source = \"/Users/elodiehusson/Desktop/AD & PD/77k_AD_PD_CTRL.h5ad\",\n",
    "    dest = \"h5seurat\",\n",
    "    assay = \"RNA\",\n",
    "    overwrite = TRUE,\n",
    "    reductions = FALSE,\n",
    "    graphs = FALSE,\n",
    "    misc = FALSE,\n",
    "    verbose = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e87afa4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "obj <- LoadH5Seurat(\n",
    "    \"/Users/elodiehusson/Desktop/AD & PD/77k_AD_PD_CTRL.h5seurat\",\n",
    "    assays = list(RNA = \"data\"),  # ne charge QUE data\n",
    "    reductions = FALSE,\n",
    "    graphs = FALSE\n",
    ")\n",
    "obj"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
