{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "26139408",
   "metadata": {},
   "source": [
    "0. Packages & options (≃ “Importing Packages”)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "be15ea8a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "the.seed <-42\n",
    "set.seed(the.seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1607d770",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//Rtmpwfkr2C/downloaded_packages\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//Rtmpwfkr2C/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GitHub PAT from the git credential store.\n",
      "\n",
      "Downloading GitHub repo mojaveazure/seurat-disk@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "digest       (0.6.37   -> 0.6.39  ) [CRAN]\n",
      "listenv      (0.9.1    -> 0.10.0  ) [CRAN]\n",
      "future       (1.67.0   -> 1.68.0  ) [CRAN]\n",
      "spatstat.... (3.1-4    -> 3.1-5   ) [CRAN]\n",
      "spatstat.... (3.4-2    -> 3.4-3   ) [CRAN]\n",
      "spatstat.... (3.6-0    -> 3.6-1   ) [CRAN]\n",
      "promises     (1.4.0    -> 1.5.0   ) [CRAN]\n",
      "stringr      (1.5.2    -> 1.6.0   ) [CRAN]\n",
      "S7           (0.2.0    -> 0.2.1   ) [CRAN]\n",
      "RcppArmad... (15.0.2-2 -> 15.2.2-1) [CRAN]\n",
      "reshape2     (1.4.4    -> 1.4.5   ) [CRAN]\n",
      "ggplot2      (4.0.0    -> 4.0.1   ) [CRAN]\n",
      "tinytex      (0.57     -> 0.58    ) [CRAN]\n",
      "yaml         (2.3.10   -> 2.3.11  ) [CRAN]\n",
      "purrr        (1.1.0    -> 1.2.0   ) [CRAN]\n",
      "progressr    (0.17.0   -> 0.18.0  ) [CRAN]\n",
      "uwot         (0.2.3    -> 0.2.4   ) [CRAN]\n",
      "spatstat.... (3.5-3    -> 3.6-0   ) [CRAN]\n",
      "reticulate   (1.44.0   -> 1.44.1  ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing 19 packages: digest, listenv, future, spatstat.univar, spatstat.random, spatstat.geom, promises, stringr, S7, RcppArmadillo, reshape2, ggplot2, tinytex, yaml, purrr, progressr, uwot, spatstat.explore, reticulate\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//Rtmpwfkr2C/downloaded_packages\n",
      "\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n",
      "* checking for file ‘/private/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T/Rtmpwfkr2C/remotes6dac7e8bd960/mojaveazure-seurat-disk-877d4e1/DESCRIPTION’ ... OK\n",
      "* preparing ‘SeuratDisk’:\n",
      "* checking DESCRIPTION meta-information ... OK\n",
      "* checking for LF line-endings in source and make files and shell scripts\n",
      "* checking for empty or unneeded directories\n",
      "Omitted ‘LazyData’ from DESCRIPTION\n",
      "* building ‘SeuratDisk_0.0.0.9021.tar.gz’\n",
      "\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/k4/d0pwylgd1tl4mf2_hqnm6l400000gn/T//Rtmpwfkr2C/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘SeuratDisk’ is not available as a binary package for this version of R\n",
      "\n",
      "A version of this package for your version of R might be available elsewhere,\n",
      "see the ideas at\n",
      "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”\n",
      "Warning message:\n",
      "“package ‘SeuratDisk’ is not available for this version of R\n",
      "\n",
      "A version of this package for your version of R might be available elsewhere,\n",
      "see the ideas at\n",
      "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”\n",
      "Warning message:\n",
      "“package ‘SeuratDisk’ is not available as a binary package for this version of R\n",
      "\n",
      "A version of this package for your version of R might be available elsewhere,\n",
      "see the ideas at\n",
      "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"Seurat\")\n",
    "install.packages(\"remotes\")\n",
    "remotes::install_github(\"mojaveazure/seurat-disk\")\n",
    "\n",
    "install.packages(\"hdf5r\", type = \"binary\")\n",
    "install.packages(\"SeuratDisk\", type = \"binary\")\n",
    "install.packages(\"SeuratDisk\")\n",
    "install.packages(\"SeuratDisk\", type = \"binary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "29980d13",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘ggplot2’ was built under R version 4.5.2”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in value[[3L]](cond): Package ‘ggplot2’ version 4.0.0 cannot be unloaded:\n Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘cowplot’, ‘ggridges’, ‘scattermore’, ‘ggrepel’, ‘Seurat’, ‘patchwork’, ‘plotly’, ‘sctransform’ so cannot be unloaded\n\n",
     "output_type": "error",
     "traceback": [
      "Error in value[[3L]](cond): Package ‘ggplot2’ version 4.0.0 cannot be unloaded:\n Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘cowplot’, ‘ggridges’, ‘scattermore’, ‘ggrepel’, ‘Seurat’, ‘patchwork’, ‘plotly’, ‘sctransform’ so cannot be unloaded\n\nTraceback:\n",
      "1. tryCatch(unloadNamespace(package), error = function(e) {\n .     P <- if (!is.null(cc <- conditionCall(e))) \n .         paste(\"Error in\", deparse(cc)[1L], \": \")\n .     else \"Error : \"\n .     stop(gettextf(\"Package %s version %s cannot be unloaded:\\n %s\", \n .         sQuote(package), oldversion, paste0(P, conditionMessage(e), \n .             \"\\n\")), domain = NA)\n . })",
      "2. tryCatchList(expr, classes, parentenv, handlers)",
      "3. tryCatchOne(expr, names, parentenv, handlers[[1L]])",
      "4. value[[3L]](cond)",
      "5. stop(gettextf(\"Package %s version %s cannot be unloaded:\\n %s\", \n .     sQuote(package), oldversion, paste0(P, conditionMessage(e), \n .         \"\\n\")), domain = NA)",
      "6. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = NULL)\n . }, \"Package ‘ggplot2’ version 4.0.0 cannot be unloaded:\\n Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘cowplot’, ‘ggridges’, ‘scattermore’, ‘ggrepel’, ‘Seurat’, ‘patchwork’, ‘plotly’, ‘sctransform’ so cannot be unloaded\\n\", \n .     base::quote(value[[3L]](cond)))"
     ]
    }
   ],
   "source": [
    "## Packages \n",
    "library(Seurat)\n",
    "library(SeuratDisk)   \n",
    "library(Matrix)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8a2cd58",
   "metadata": {},
   "source": [
    "1. Import des données (≃ adata = sc.read_h5ad(...))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4c9ff384",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown file type: h5ad”\n",
      "Warning message:\n",
      "“'assay' not set, setting to 'RNA'”\n",
      "Creating h5Seurat file for version 3.1.5.9900\n",
      "\n",
      "Adding X as data\n",
      "\n",
      "Adding X as counts\n",
      "\n",
      "Adding meta.features from var\n",
      "\n",
      "Adding X_umap as cell embeddings for umap\n",
      "\n",
      "Adding batch_condition to miscellaneous data\n",
      "\n",
      "Adding citation to miscellaneous data\n",
      "\n",
      "Adding class_colors to miscellaneous data\n",
      "\n",
      "Adding genome to miscellaneous data\n",
      "\n",
      "Adding organism to miscellaneous data\n",
      "\n",
      "Adding organism_ontology_term_id to miscellaneous data\n",
      "\n",
      "Adding schema_reference to miscellaneous data\n",
      "\n",
      "Adding schema_version to miscellaneous data\n",
      "\n",
      "Adding subclass_colors to miscellaneous data\n",
      "\n",
      "Adding subtype_colors to miscellaneous data\n",
      "\n",
      "Adding title to miscellaneous data\n",
      "\n",
      "Adding uid to miscellaneous data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## 1.1 Conversion h5ad -> h5seurat (à faire UNE FOIS) \n",
    "\n",
    "#elodie \n",
    "Convert(\"/Users/elodiehusson/Desktop/AD & PD/AD_PD_CTRL.h5ad\", dest = \"h5seurat\", overwrite = TRUE)\n",
    "\n",
    "#Yara\n",
    "# Convert(\"C:/Users/yarad/Desktop/x/Masters/Master BMC - Sorbonne/M2/Single Cell/Project/Coding Project/AD_PD_CTRL.h5ad\", dest = \"h5seurat\", overwrite = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ece2dc8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## 1.2 Chargement dans un objet Seurat ----\n",
    "sobj <- LoadH5Seurat(\"AD_PD_CTRL.h5seurat\")\n",
    "\n",
    "sobj\n",
    "head(colnames(sobj))      # noms des cellules\n",
    "head(rownames(sobj))      # noms des gènes\n",
    "head(sobj@meta.data)      # équivalent à adata.obs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "270f1382",
   "metadata": {},
   "source": [
    "\n",
    "2. QC & métadonnées (≃ sc.pp.calculate_qc_metrics + plots)\n",
    "\n",
    "En Python tu as :\n",
    "\n",
    "sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "\n",
    "En R/Seurat, on utilise déjà nFeature_RNA, nCount_RNA et on ajoute le % mitochondrial :\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee9832a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## 2.1 Ajout des métriques QC ----\n",
    "\n",
    "# gènes mitochondriaux humains typiques\n",
    "sobj[[\"percent.mt\"]] <- PercentageFeatureSet(sobj, pattern = \"^MT-\")\n",
    "\n",
    "# si besoin : ribosomal, hemoglobin, etc. comme dans le .Rmd\n",
    "sobj[[\"percent.ribo\"]] <- PercentageFeatureSet(sobj, pattern = \"^RPL|^RPS\")\n",
    "sobj[[\"percent.hb\"]]   <- PercentageFeatureSet(sobj, pattern = \"^HB[AB]\")\n",
    "\n",
    "head(sobj@meta.data)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59dcf93b",
   "metadata": {},
   "source": [
    "\n",
    "2.2 Visualisation des QC metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9231f42a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "(≃ sc.pl.violin(adata, ['n_counts','n_genes']) + scatter)\n",
    "\n",
    "## Violin plots nFeature, nCount, percent.mt ----\n",
    "VlnPlot(\n",
    "  sobj,\n",
    "  features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"),\n",
    "  ncol = 3,\n",
    "  pt.size = 0.1\n",
    ")\n",
    "\n",
    "## Scatter plots (équivalent des scatter QC dans scanpy) ----\n",
    "FeatureScatter(sobj, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\") +\n",
    "FeatureScatter(sobj, feature1 = \"nCount_RNA\", feature2 = \"percent.mt\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7ff8a63",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "3. Filtrage des cellules (≃ “Filtering low quality cells / mt / ribo / hb”)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba1c465",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## 3.1 Choix des thresholds (inspirés du .Rmd + ton notebook) ----\n",
    "minGenes <- 200      # comme dans la prof\n",
    "minUMI   <- 1000     # à adapter\n",
    "maxMT    <- 15       # % de mitochondrie max (exemple)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92a5ac00",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## 3.2 Création d'un objet filtré ----\n",
    "sobj_filtrd <- subset(\n",
    "  sobj,\n",
    "  subset = nFeature_RNA > minGenes &\n",
    "           nCount_RNA   > minUMI   &\n",
    "           percent.mt   < maxMT\n",
    ")\n",
    "\n",
    "## Visualisation post-filtre ----\n",
    "VlnPlot(\n",
    "  sobj_filtrd,\n",
    "  features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"),\n",
    "  ncol = 3,\n",
    "  pt.size = 0.1\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74891600",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#Si tu veux exclure certains gènes (ribosomal, hémoglobine) comme mentionné dans ton notebook :\n",
    "\n",
    "genes_to_remove <- grep(\"^RPL|^RPS|^HB[AB]\", rownames(sobj_filtrd), value = TRUE)\n",
    "\n",
    "sobj_filtrd <- sobj_filtrd[ setdiff(rownames(sobj_filtrd), genes_to_remove), ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db3b148b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "4. Normalisation (≃ sc.pp.normalize_total + sc.pp.log1p)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4820f89",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "\n",
    "5. HVG – Highly Variable Genes (≃ sc.pp.highly_variable_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98a5f4ac",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "sobj_filtrd <- FindVariableFeatures(\n",
    "  sobj_filtrd,\n",
    "  selection.method = \"vst\",\n",
    "  nfeatures = 3000\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91083f24",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# liste des HVG\n",
    "hvg <- VariableFeatures(sobj_filtrd)\n",
    "length(hvg)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e4aa172",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "6. Scaling & PCA (≃ sc.pp.scale + sc.tl.pca)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72f89fca",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sobj_filtrd <- ScaleData(sobj_filtrd, features = hvg)\n",
    "\n",
    "sobj_filtrd <- RunPCA(\n",
    "  sobj_filtrd,\n",
    "  features = hvg,\n",
    "  npcs = 50 # à adapter selon ton notebook\n",
    ")\n",
    "\n",
    "ElbowPlot(sobj_filtrd)  # pour choisir combien de PC garder\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c81a373a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "7. Voisins, clustering, UMAP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541ec667",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#7.1 Graph des voisins\n",
    "\n",
    "sobj_filtrd <- FindNeighbors(\n",
    "  sobj_filtrd,\n",
    "  dims = 1:30   # adapte au nb de PC retenues\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eddd9ee8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#7.2 Clustering Leiden ≃ Seurat FindClusters\n",
    "\n",
    "sobj_filtrd <- FindClusters(\n",
    "  sobj_filtrd,\n",
    "  resolution = 0.5  # adapte à ce que tu as testé en Python\n",
    ")\n",
    "\n",
    "table(sobj_filtrd$seurat_clusters)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae31e5c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#7.3 UMAP\n",
    "\n",
    "sobj_filtrd <- RunUMAP(\n",
    "  sobj_filtrd,\n",
    "  dims = 1:30\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48cbc062",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "DimPlot(sobj_filtrd, reduction = \"umap\", group.by = \"seurat_clusters\", label = TRUE)\n",
    "DimPlot(sobj_filtrd, reduction = \"umap\", group.by = \"disease\")\n",
    "DimPlot(sobj_filtrd, reduction = \"umap\", group.by = \"genetic_ancestry\")\n",
    "DimPlot(sobj_filtrd, reduction = \"umap\", group.by = \"subtype\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b96fea",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "8. Annotation manuelle (≃ section “Manual Annotation” + marker plots)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b0c773",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Exemple de gènes marqueurs (adaptés à ton dataset) ----\n",
    "marker_genes <- c(\"CD3D\", \"CD4\", \"CD8A\", \"P2RY12\", \"GFAP\")  # à adapter\n",
    "\n",
    "# UMAP colored by expression\n",
    "FeaturePlot(sobj_filtrd, features = marker_genes, cols = c(\"lightgrey\", \"red\"))\n",
    "\n",
    "# Violin plots par cluster\n",
    "VlnPlot(sobj_filtrd, features = marker_genes, group.by = \"seurat_clusters\", pt.size = 0)\n",
    "\n",
    "Ensuite tu peux renommer tes clusters (comme dans le .Rmd de la prof) :\n",
    "\n",
    "new.cluster.ids <- c(\n",
    "  \"0\" = \"CD8_T\",\n",
    "  \"1\" = \"CD4_T\",\n",
    "  \"2\" = \"Microglia\",\n",
    "  \"3\" = \"Astrocytes\"\n",
    "  # etc.\n",
    ")\n",
    "\n",
    "sobj_filtrd$celltype_manual <- plyr::mapvalues(\n",
    "  x = sobj_filtrd$seurat_clusters,\n",
    "  from = names(new.cluster.ids),\n",
    "  to   = new.cluster.ids\n",
    ")\n",
    "\n",
    "DimPlot(sobj_filtrd, reduction = \"umap\", group.by = \"celltype_manual\", label = TRUE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da567e84",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "9. Annotation automatique (≃ CellTypist)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64c57176",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# En Python tu utilises celltypist.\n",
    "# En R, il n’y a pas CellTypist directement, mais tu peux utiliser par ex. SingleR ou scType.\n",
    "\n",
    "# Pour rester simple (et puisque ce n’est pas dans le .Rmd prof), je te mets juste un squelette très court :\n",
    "\n",
    "# Exemple *idée* avec SingleR (si tu veux pousser) :\n",
    "# library(SingleR)\n",
    "# library(celldex)\n",
    "# ref <- HumanPrimaryCellAtlasData()\n",
    "# sce <- as.SingleCellExperiment(sobj_filtrd)\n",
    "# pred <- SingleR(test = sce, ref = ref, labels = ref$label.main)\n",
    "# sobj_filtrd$celltype_auto <- pred$labels\n",
    "# DimPlot(sobj_filtrd, group.by = \"celltype_auto\", label = TRUE)\n",
    "\n",
    "# Tu peux aussi t’arrêter à l’annotation manuelle.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05c1c2eb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "10. Differential Expression (≃ sc.tl.rank_genes_groups)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a52e024c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## 10.1 Choisir l'identité par laquelle comparer ----\n",
    "Idents(sobj_filtrd) <- \"disease\"\n",
    "\n",
    "## 10.2 Comparer un groupe vs référence ----\n",
    "markers_AD_vs_ctrl <- FindMarkers(\n",
    "  sobj_filtrd,\n",
    "  ident.1 = \"dementia || Alzheimer disease\",  # adapte au label exact\n",
    "  ident.2 = \"normal\",\n",
    "  logfc.threshold = 0.25,\n",
    "  min.pct = 0.2,\n",
    "  test.use = \"wilcox\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cc432c7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "head(markers_AD_vs_ctrl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2296ff9a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "Idents(sobj_filtrd) <- \"seurat_clusters\"\n",
    "\n",
    "markers_all <- FindAllMarkers(\n",
    "  sobj_filtrd,\n",
    "  only.pos = TRUE,\n",
    "  min.pct = 0.25,\n",
    "  logfc.threshold = 0.25\n",
    ")\n",
    "\n",
    "head(markers_all)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e39b2505",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "11. Pseudo-bulk (≃ adata.to_df().groupby(adata.obs[\"donor_id\"]).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "611792d5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "## 11.1 Récupérer la matrice de counts ----\n",
    "counts <- GetAssayData(sobj_filtrd, slot = \"counts\")\n",
    "\n",
    "## 11.2 Vérifier qu'on a bien une colonne donor_id dans la metadata ----\n",
    "head(sobj_filtrd$donor_id)\n",
    "\n",
    "## 11.3 Agrégation pseudo-bulk par donneur ----\n",
    "donors <- sobj_filtrd$donor_id\n",
    "\n",
    "pb_mat <- sapply(\n",
    "  X = split(seq_len(ncol(counts)), donors),\n",
    "  FUN = function(cells) Matrix::rowSums(counts[, cells, drop = FALSE])\n",
    ")\n",
    "\n",
    "pb_mat <- as.matrix(pb_mat)  # gènes x donneurs\n",
    "\n",
    "dim(pb_mat)\n",
    "colnames(pb_mat)  # donneurs\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
